### DISCUSSION TITLE
Possible `nb::init_implicit` breakage on MSVC due to initializer lists

### DISCUSSION THREAD

**Discussion by nicholasjng** [wjakob/nanobind#122] [Q&A] [answered]
Hey! I am back with another question. I am porting over a C++ class from pybind wrapping a `double`, with a single constructor like so (more or less):

```cpp
class Counter {
    public:
        ...
        double value;
        Counter(double v = 0.): value(v) {};
```

In pybind, this was bound to Python with `py::implicitly_convertible<py::float_, Counter>()`, `py::implicitly_convertible<py::int_, Counter>()`, to enable implicit construction from ints and floats directly in Python.

Since there is no `nb::float_`, I opted to go for `.def(nb::init_implicit<double>)` instead, which works fine on Linux and Mac, but crashes on Windows/MSVC with an error about narrowing conversions (C3298):

```
error C2398: Element '1': conversion from 'Arg' to 'double' requires a narrowing conversion
          with
          [
              Arg=nanobind::int_
          ]
  C:\users\runneradmin\_bazel_runneradmin\zjvmdg72\execroot\com_github_google_benchmark\external\nanobind\include\nanobind\nb_class.h(328): note: see reference to function template instantiation 'void nanobind::detail::init_implicit<nanobind::int_>::execute<nanobind::class_<benchmark::Counter>,>(Class &)' being compiled
          with
          [
              Class=nanobind::class_<benchmark::Counter>
          ]
```

I did some googling and found out that this happens in initializer lists, and `init_implicit` uses one in here: 
https://github.com/wjakob/nanobind/blob/fe3ecb800a7a3e8023e8ee77167a6241591e0b8b/include/nanobind/nb_class.h#L182-L186 

Is there a way I can still enable implicit construction from both floats and ints in this case? I tried binding a `.def(__init__, ...)` for an int value by hand which instantiates the `Counter` from a `static_cast<double>(v)`, but to no avail.

**Comment by wjakob** [comment, marked as answer]
This only works for constructors that actually exist in your type. You will need a *placement new* constructor as described in `README.md`, along with `nb::implicitly_convertible<float, Counter>()`.

**Reply by nicholasjng** [reply]
Interesting, I understood the readme comment as "`nb::init_implicit` has replaced `py::implicitly_convertible`". Thanks for the clarification.

So if I want to implicitly construct my Python `Counter` from a Python `int`, but my C++ `Counter` only has a double constructor, I need to use `nb::implicitly_convertible` as you mentioned?

**Reply by wjakob** [reply]
yep.
